File: apps/chandao-cli/src/actions/story.rs
use clap::Subcommand;
use serde_json::json;
use std::cell::RefCell;
use std::rc::Rc;

use crate::auth::AuthManager;
use crate::client::{AuthenticatedClient, Client};
use crate::markdown::markdown_to_html;
use crate::utils;

macro_rules! with_auth {
    ($client:expr, $auth:expr, $body:expr) => {{
        let mut ac = $client.authenticate($auth)?;
        $body(&mut ac)
    }};
}

#[derive(Subcommand)]
pub enum StoryCommands {
    /// List stories
    List {
        /// Product ID
        #[arg(short = 'p', long)]
        product: Option<i64>,
        /// Project ID
        #[arg(short = 'j', long)]
        project: Option<i64>,
        /// Execution ID
        #[arg(short = 'e', long)]
        execution: Option<i64>,
        /// Browse type (all/unclosed/bysearch)
        #[arg(short = 'b', long, default_value = "all")]
        browse_type: String,
        /// Order by
        #[arg(short = 'o', long, default_value = "id_desc")]
        order_by: String,
        /// Page
        #[arg(short = 'g', long, default_value = "1")]
        page: u32,
        /// Limit
        #[arg(short = 'n', long, default_value = "20")]
        limit: u32,
    },
    /// Get story
    Get { id: i64 },
    /// Create story
    Create {
        /// Product ID
        #[arg(short = 'p', long)]
        product: i64,
        /// Title
        #[arg(short, long)]
        title: String,
        /// Specification
        #[arg(short = 'S', long)]
        spec: Option<String>,
        /// Verification criteria
        #[arg(long)]
        verify: Option<String>,
        /// Module ID
        #[arg(short = 'm', long)]
        module: Option<i64>,
        /// Parent story ID
        #[arg(long)]
        parent: Option<i64>,
        /// Priority (1-4)
        #[arg(short = 'i', long)]
        pri: Option<u8>,
        /// Category (feature/interface/performance/safe/experience/improve/other)
        #[arg(short = 'c', long)]
        category: Option<String>,
        /// Source
        #[arg(short = 'o', long)]
        source: Option<String>,
        /// Assigned to
        #[arg(short = 'a', long)]
        assigned: Option<String>,
        /// Reviewer
        #[arg(short = 'r', long)]
        reviewer: Option<String>,
        /// Estimate (hours)
        #[arg(long)]
        estimate: Option<f64>,
        /// Project ID
        #[arg(short = 'j', long)]
        project: Option<i64>,
        /// Execution ID
        #[arg(short = 'e', long)]
        execution: Option<i64>,
        /// Dry run
        #[arg(long)]
        dry_run: bool,
    },
    /// Update story
    Update {
        /// Story ID
        id: i64,
        #[arg(short, long)]
        title: Option<String>,
        /// Specification
        #[arg(short = 'S', long)]
        spec: Option<String>,
        #[arg(short = 'm', long)]
        module: Option<i64>,
        /// Parent story ID
        #[arg(long)]
        parent: Option<i64>,
        #[arg(short = 'i', long)]
        pri: Option<u8>,
        /// Category (feature/interface/performance/safe/experience/improve/other)
        #[arg(short = 'c', long)]
        category: Option<String>,
        /// Source
        #[arg(short = 'o', long)]
        source: Option<String>,
        #[arg(short = 'a', long)]
        assigned: Option<String>,
        #[arg(short, long)]
        status: Option<String>,
        #[arg(long)]
        dry_run: bool,
    },
    /// Review a story (approve/reject)
    Review {
        /// Story ID
        id: i64,
        /// Result: pass, reject, revert
        #[arg(short, long)]
        result: String,
        /// Review comment
        #[arg(short, long)]
        comment: Option<String>,
        #[arg(long)]
        dry_run: bool,
    },
    /// Close a story
    Close {
        id: i64,
        /// Reason: done, subdivided, duplicate, postponed, willnotdo, cancel, bydesign
        #[arg(short, long)]
        reason: String,
        #[arg(long)]
        dry_run: bool,
    },
    /// Activate a closed story
    Activate {
        id: i64,
        /// Assigned to
        #[arg(short = 'a', long)]
        assigned: Option<String>,
        /// Comment
        #[arg(short, long)]
        comment: Option<String>,
        #[arg(long)]
        dry_run: bool,
    },
    /// Delete story
    Delete {
        id: i64,
        #[arg(long)]
        yes: bool,
        #[arg(long)]
        dry_run: bool,
    },
    /// Change a story (submit for review)
    Change {
        /// Story ID
        id: i64,
        /// Reviewers (comma-separated)
        #[arg(short, long, value_delimiter = ',', required = true)]
        reviewer: Vec<String>,
        /// New title
        #[arg(short, long)]
        title: Option<String>,
        /// New specification
        #[arg(long)]
        spec: Option<String>,
        /// New verification criteria
        #[arg(long)]
        verify: Option<String>,
        #[arg(long)]
        dry_run: bool,
    },
}

pub fn handle_story(
    client: &Client,
    auth: &Rc<RefCell<AuthManager>>,
    cmd: &StoryCommands,
) -> Result<(), String> {
    match cmd {
        StoryCommands::List { product, project, execution, browse_type, order_by, page, limit } => with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
            let path = if let Some(e) = execution {
                format!("/executions/{}/stories?browseType={}&orderBy={}&pageID={}&recPerPage={}", e, browse_type, order_by, page, limit)
            } else if let Some(p) = project {
                format!("/projects/{}/stories?browseType={}&orderBy={}&pageID={}&recPerPage={}", p, browse_type, order_by, page, limit)
            } else if let Some(p) = product {
                format!("/products/{}/stories?browseType={}&orderBy={}&pageID={}&recPerPage={}", p, browse_type, order_by, page, limit)
            } else {
                format!("/stories?browseType={}&orderBy={}&pageID={}&recPerPage={}", browse_type, order_by, page, limit)
            };
            let data = ac.get(&path)?;
            utils::print_table(&data, &["id", "title", "status", "pri", "productName", "openedDate"]);
            Ok(())
        }),
        StoryCommands::Get { id } => with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
            let data = ac.get(&format!("/stories/{}", id))?;
            utils::print_json(&data);
            Ok(())
        }),
        StoryCommands::Create { product, title, spec, verify, module, parent, pri, category, source, assigned, reviewer, estimate, project, execution, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 创建需求: {}", title); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({"productID": product, "title": title});
                if let Some(d) = spec { body["spec"] = json!(markdown_to_html(&d)); }
                if let Some(v) = verify { body["verify"] = json!(markdown_to_html(&v)); }
                if let Some(m) = module { body["module"] = json!(m); }
                if let Some(pa) = parent { body["parent"] = json!(pa); }
                if let Some(p) = pri { body["pri"] = json!(p); }
                if let Some(c) = category { body["category"] = json!(c); }
                if let Some(s) = source { body["source"] = json!(s); }
                if let Some(a) = assigned { body["assignedTo"] = json!(a); }
                if let Some(e) = estimate { body["estimate"] = json!(e); }
                if let Some(j) = project { body["project"] = json!(j); }
                if let Some(ex) = execution { body["execution"] = json!(ex); }
                if let Some(r) = reviewer { body["reviewer"] = json!(r); }
                let result = ac.post("/stories", &body)?;
                println!("✅ 需求创建成功");
                utils::print_json(&result);
                Ok(())
            })
        }
        StoryCommands::Update { id, title, spec, module, parent, pri, category, source, assigned, status, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 更新需求 #{}", id); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({});
                if let Some(t) = title { body["title"] = json!(t); }
                if let Some(s) = spec { body["spec"] = json!(markdown_to_html(&s)); }
                if let Some(m) = module { body["module"] = json!(m); }
                if let Some(pa) = parent { body["parent"] = json!(pa); }
                if let Some(p) = pri { body["pri"] = json!(p); }
                if let Some(c) = category { body["category"] = json!(c); }
                if let Some(s) = source { body["source"] = json!(s); }
                if let Some(a) = assigned { body["assignedTo"] = json!(a); }
                if let Some(s) = status { body["status"] = json!(s); }
                let result = ac.put(&format!("/stories/{}", id), &body)?;
                println!("✅ 需求 #{} 更新成功", id);
                utils::print_json(&result);
                Ok(())
            })
        }
        StoryCommands::Review { id, result, comment, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 评审需求 #{}: {}", id, result); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({"result": result});
                if let Some(c) = comment { body["comment"] = json!(c); }
                ac.post(&format!("/stories/{}/review", id), &body)?;
                println!("✅ 需求 #{} 评审完成: {}", id, result);
                Ok(())
            })
        }
        StoryCommands::Close { id, reason, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 关闭需求 #{}", id); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let body = json!({"closedReason": reason});
                ac.put(&format!("/stories/{}/close", id), &body)?;
                println!("✅ 需求 #{} 已关闭", id);
                Ok(())
            })
        }
        StoryCommands::Activate { id, assigned, comment, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 激活需求 #{}", id); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({});
                if let Some(a) = assigned { body["assignedTo"] = json!(a); }
                if let Some(c) = comment { body["comment"] = json!(c); }
                ac.put(&format!("/stories/{}/activate", id), &body)?;
                println!("✅ 需求 #{} 已激活", id);
                Ok(())
            })
        }
        StoryCommands::Delete { id, yes, dry_run } => {
            if !yes && !dry_run { eprintln!("⚠️  确认删除需求 #{}？使用 --yes", id); return Ok(()); }
            if *dry_run { println!("🔍 [DRY-RUN] 删除需求 #{}", id); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                ac.delete(&format!("/stories/{}", id))?;
                println!("✅ 需求 #{} 已删除", id);
                Ok(())
            })
        }
        StoryCommands::Change { id, reviewer, title, spec, verify, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 变更需求 #{}: reviewer={:?}", id, reviewer); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({"reviewer": reviewer});
                if let Some(t) = title { body["title"] = json!(t); }
                if let Some(s) = spec { body["spec"] = json!(markdown_to_html(&s)); }
                if let Some(v) = verify { body["verify"] = json!(markdown_to_html(&v)); }
                let result = ac.put(&format!("/stories/{}/change", id), &body)?;
                println!("✅ 需求 #{} 已提交变更", id);
                utils::print_json(&result);
                Ok(())
            })
        }
    }
}

File: skills/chandao/scripts/story.js
#!/usr/bin/env node
/**
 * chandao story (requirement) module
 *
 * Usage:
 *   node story.js --action list [--product N] [--project N] [--execution N]
 *   node story.js --action get --id <id>
 *   node story.js --action create --product <id> --title <title>
 *   node story.js --action update --id <id> --title <title>
 *   node story.js --action change --id <id> --reviewer <a1,a2>
 *   node story.js --action close --id <id> --reason <done|subdivided|duplicate|postponed|willnotdo|cancel|bydesign>
 *   node story.js --action activate --id <id>
 *   node story.js --action delete --id <id>
 *   node story.js --action <write-action> --dry-run
 *   node story.js --action delete --yes
 */

import { api, paginate } from "./auth.js";

const WRITE_ACTIONS = new Set([
  "create",
  "update",
  "change",
  "close",
  "activate",
  "delete",
]);

function parseArgs() {
  const args = process.argv.slice(2);
  const params = {};
  for (let i = 0; i < args.length; i++) {
    if (args[i].startsWith("--")) {
      const key = args[i].slice(2);
      params[key] = args[i + 1] ?? true;
      if (params[key] !== true) i++;
    } else if (!params.action) {
      params.action = args[i];
    }
  }
  if (typeof params.reviewer === "string") {
    params.reviewer = params.reviewer.split(",");
  }
  return params;
}

function checkDryRun(action) {
  if (WRITE_ACTIONS.has(action) && process.argv.includes("--dry-run")) {
    console.log("[DRY RUN] Would execute:", process.argv.slice(2).join(" "));
    process.exit(0);
  }
}

function checkYes(action) {
  if (action === "delete" && !process.argv.includes("--yes")) {
    console.log("Delete requires --yes to confirm.");
    process.exit(0);
  }
}

// ── Actions ───────────────────────────────────────────────────────
async function list(params) {
  let path = "/stories";
  if (params.product) path = `/products/${params.product}/stories`;
  else if (params.project) path = `/projects/${params.project}/stories`;
  else if (params.execution) path = `/executions/${params.execution}/stories`;

  path += paginate("", {
    limit: params.limit,
    page: params.page,
    query: { orderBy: "id_desc" },
  });

  const { data } = await api("GET", path);
  const items = data.stories || data.data || [];
  if (!items.length) {
    console.log("No stories found.");
    return;
  }

  const header = `| id | title | status | pri |`;
  console.log(header);
  console.log("─".repeat(header.length));
  for (const s of items) {
    console.log(
      `| ${s.id} | ${(s.title || "").slice(0, 40)} | ${s.status || "-"} | ${s.pri || "-"} |`,
    );
  }
  console.log(`\nTotal: ${items.length}`);
}

async function get(params) {
  const { data } = await api("GET", `/stories/${params.id}`);
  console.log(JSON.stringify(data, null, 2));
}

async function create(params) {
  if (!params.product || !params.title) {
    console.error("Error: create requires --product and --title");
    process.exit(1);
  }

  const body = {
    productID: Number(params.product),
    title: params.title,
  };
  if (params.pri) body.pri = Number(params.pri);
  if (params.module) body.module = Number(params.module);
  if (params.parent) body.parent = Number(params.parent);
  if (params.estimate) body.estimate = Number(params.estimate);
  if (params.spec) body.spec = params.spec;
  if (params.category) body.category = params.category;
  if (params.source) body.source = params.source;
  if (params.verify) body.verify = params.verify;
  if (params.assignedTo) body.assignedTo = params.assignedTo;
  if (params.reviewer)
    body.reviewer =
      typeof params.reviewer === "string"
        ? params.reviewer.split(",")
        : params.reviewer;
  if (params.project) body.project = Number(params.project);
  if (params.execution) body.execution = Number(params.execution);

  const { data } = await api("POST", "/stories", body);
  if (data.status === "success") {
    console.log(`Story created successfully. ID: ${data.id}`);
  } else {
    console.error("Failed to create story:", JSON.stringify(data));
    process.exit(1);
  }
}

async function update(params) {
  if (!params.id) {
    console.error("Error: update requires --id");
    process.exit(1);
  }

  const { data: current } = await api("GET", `/stories/${params.id}`);
  if (current.status !== "success") {
    console.error("Failed to get story:", JSON.stringify(current));
    process.exit(1);
  }

  const body = {};
  const fields = [
    "title",
    "spec",
    "verify",
    "module",
    "parent",
    "pri",
    "category",
    "source",
    "assignedTo",
    "product",
  ];
  for (const f of fields) {
    if (params[f] !== undefined) body[f] = params[f];
    else if (current[f] !== undefined) body[f] = current[f];
  }
  // Only update status if explicitly provided
  if (params.status) body.status = params.status;

  const { data } = await api("PUT", `/stories/${params.id}`, body);
  if (data.status === "success") {
    console.log(`Story ${params.id} updated successfully.`);
  } else {
    console.error("Failed to update story:", JSON.stringify(data));
    process.exit(1);
  }
}

async function change(params) {
  if (!params.id) {
    console.error("Error: change requires --id");
    process.exit(1);
  }
  if (!params.reviewer) {
    console.error("Error: change requires --reviewer");
    process.exit(1);
  }

  const reviewer =
    typeof params.reviewer === "string"
      ? params.reviewer.split(",")
      : params.reviewer;
  const body = { reviewer };
  if (params.title) body.title = params.title;
  if (params.spec) body.spec = params.spec;
  if (params.verify) body.verify = params.verify;

  const { data } = await api("PUT", `/stories/${params.id}/change`, body);
  if (data.status === "success") {
    console.log(`Story ${params.id} change submitted for review.`);
  } else {
    console.error("Failed to change story:", JSON.stringify(data));
    process.exit(1);
  }
}

async function close(params) {
  if (!params.id) {
    console.error("Error: close requires --id");
    process.exit(1);
  }
  if (!params.reason) {
    console.error(
      "Error: close requires --reason (done|subdivided|duplicate|postponed|willnotdo|cancel|bydesign)",
    );
    process.exit(1);
  }

  const body = { closedReason: params.reason };
  if (params.comment) body.comment = params.comment;

  const { data } = await api("PUT", `/stories/${params.id}/close`, body);
  if (data.status === "success") {
    console.log(`Story ${params.id} closed.`);
  } else {
    console.error("Failed to close story:", JSON.stringify(data));
    process.exit(1);
  }
}

async function activate(params) {
  if (!params.id) {
    console.error("Error: activate requires --id");
    process.exit(1);
  }

  const { data } = await api("PUT", `/stories/${params.id}/activate`);
  if (data.status === "success") {
    console.log(`Story ${params.id} activated.`);
  } else {
    console.error("Failed to activate story:", JSON.stringify(data));
    process.exit(1);
  }
}

async function del(params) {
  const { data } = await api("DELETE", `/stories/${params.id}`);
  if (data.status === "success") {
    console.log(`Story ${params.id} deleted.`);
  } else {
    console.error("Failed to delete story:", JSON.stringify(data));
    process.exit(1);
  }
}

// ── Dispatch ──────────────────────────────────────────────────────
const actions = {
  list,
  get,
  create,
  update,
  change,
  close,
  activate,
  delete: del,
};

async function main() {
  const params = parseArgs();
  if (!params.action || !actions[params.action]) {
    console.error(
      `Usage: node story.js --action <${Object.keys(actions).join("|")}>`,
    );
    process.exit(1);
  }

  checkDryRun(params.action);
  checkYes(params.action);

  await actions[params.action](params);
}

main().catch((err) => {
  console.error("Fatal error:", err.message);
  process.exit(1);
});
